packages/extension/src/serviceWorker/chromeListeners/runtimeOnMessageListener.ts

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import synchnorousCookieStore from '../../store/synchnorousCookieStore';
+import {
+  DEVTOOLS_CLOSE,
+  DEVTOOLS_OPEN,
+  DEVTOOLS_SET_JAVASCSCRIPT_COOKIE,
+  INITIAL_SYNC,
+  POPUP_CLOSE,
+  POPUP_OPEN,
+  SERVICE_WORKER_RELOAD_MESSAGE,
+  SERVICE_WORKER_TABS_RELOAD_COMMAND,
+  SET_TAB_TO_READ,
+} from '../../constants';
+import listenToNewTab from '../../utils/listenToNewTab';
+import attachCDP from '../attachCDP';
+import reloadCurrentTab from '../../utils/reloadCurrentTab';
+import { getTab } from '../../utils/getTab';
+import sendMessageWrapper from '../../utils/sendMessageWrapper';
+
+// eslint-disable-next-line complexity
+export const runtimeOnMessageListener = async (request: any) => {
+  if (!request.type) {
+    return;
+  }
+
+  const incomingMessageType = request.type;
+
+  if (SET_TAB_TO_READ === incomingMessageType) {
+    synchnorousCookieStore.tabToRead = request?.payload?.tabId?.toString();
+    const newTab = await listenToNewTab(request?.payload?.tabId);
+    // Can't use sendResponse as delay is too long. So using sendMessage instead.
+    await sendMessageWrapper(SET_TAB_TO_READ, {
+      tabId: Number(newTab),
+    });
+
+    if (synchnorousCookieStore.globalIsUsingCDP) {
+      await attachCDP({ tabId: Number(newTab) });
+    }
+
+    await reloadCurrentTab(Number(newTab));
+  }
+
+  if (SERVICE_WORKER_TABS_RELOAD_COMMAND === incomingMessageType) {
+    const sessionStorage = await chrome.storage.session.get();
+    if (Object.keys(sessionStorage).includes('allowedNumberOfTabs')) {
```

Comment:
Optional Chaining can help
```suggestion
    if (sessionStorage?.allowedNumberOfTabs) {
```